Wrong codegen for default parameter values that calls `var Object` procedure with ARC/ORC on C++ backend.

### Description

The following code compiles for C and JS backends but not for C++.

```nim
type Object = object
  foo: int

proc takeFoo(self: var Object): int =
  result = self.foo
  self.foo = 0

proc doSomething(self: var Object; foo: int = self.takeFoo()) =
  discard

proc main() =
  var obj = Object(foo: 2)
  obj.doSomething()

when isMainModule:
  main()
```

The Nim compiler uses C++ reference type as pointers, which causes a compilation error.
This problem does not exist on `refc`.

Changing the `self` parameter of the `takeFoo` proc from `var Object` to `Object` solves the problem, but in cases where `var Object` is needed, it is not a solution.

### Nim Version

Nim Compiler Version 1.9.5 [Windows: amd64]
Compiled at 2023-07-31
Copyright (c) 2006-2023 by Andreas Rumpf   

git hash: https://github.com/nim-lang/Nim/commit/281016a8022b8e8308e3e578b2c2daa6df4a66a1
active boot switches: -d:release

### Current Output

```text
C:\Users\User\nimcache\cpp_d\@mcpp.nim.cpp: In function 'void main__cpp_u9()':
C:\Users\User\nimcache\cpp_d\@mcpp.nim.cpp:122:50: error: 'T2_' declared as reference but not initialized
  122 |         tyObject_Object__yRSLqvpttiYRfCwKCKlhOA& T2_;
      |                                                  ^~~
C:\Users\User\nimcache\cpp_d\@mcpp.nim.cpp:130:15: error: invalid cast of a prvalue expression 
of type 'int' to type 'tyObject_Object__yRSLqvpttiYRfCwKCKlhOA&'
  130 |         T2_ = (tyObject_Object__yRSLqvpttiYRfCwKCKlhOA&)0;
      |               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Error: execution of an external compiler program 'g++.exe -c -std=gnu++17 -funsigned-char  -w -fmax-errors=3 -fpermissive -mno-ms-bitfields   -IC:\Users\User\.choosenim\toolchains\nim-#devel\lib -IC:\Projects\TestNim -o C:\Users\User\nimcache\cpp_d\@mcpp.nim.cpp.o C:\Users\User\nimcache\cpp_d\@mcpp.nim.cpp' failed with exit code: 1
```


### Expected Output

```text
Successful compilation for C++ backend.
```


### Possible Solution

_No response_

### Additional Information

_No response_

--- a/compiler/semexprs.nim
+++ b/compiler/semexprs.nim
@@ -2939,7 +2939,7 @@ proc hoistParamsUsedInDefault(c: PContext, call, letSection, defExpr: var PNode)
 
       call[paramPos] = newSymNode(hoistedVarSym) # Refer the original arg to its hoisted sym
 
-    # arg we refer to is a sym, wether introduced by hoisting or not doesn't matter, we simply reuse it
+    # arg we refer to is a sym, whether introduced by hoisting or not doesn't matter, we simply reuse it
     defExpr = call[paramPos]
   else:
     for i in 0..<defExpr.safeLen:

--- a/compiler/semexprs.nim
+++ b/compiler/semexprs.nim
@@ -2927,7 +2927,7 @@ proc hoistParamsUsedInDefault(c: PContext, call, letSection, defExpr: var PNode)
   if defExpr.kind == nkSym and defExpr.sym.kind == skParam and defExpr.sym.owner == call[0].sym:
     let paramPos = defExpr.sym.position + 1
 
-    if call[paramPos].kind != nkSym:
+    if call[paramPos].skipAddr.kind != nkSym:
       let hoistedVarSym = newSym(skLet, getIdent(c.graph.cache, genPrefix), c.idgen,
                                  c.p.owner, letSection.info, c.p.owner.options)
       hoistedVarSym.typ = call[paramPos].typ

--- a/tests/ccgbugs2/tcodegen.nim
+++ b/tests/ccgbugs2/tcodegen.nim
@@ -26,3 +26,22 @@ block: # NRVO2
   let innerAddress = x.innerAddress
   let outerAddress = cast[uint](x.addr)
   doAssert(innerAddress == outerAddress) # [OK]
+
+block: # bug #22354
+  type Object = object
+    foo: int
+
+  proc takeFoo(self: var Object): int =
+    result = self.foo
+    self.foo = 999
+
+  proc doSomething(self: var Object; foo: int = self.takeFoo()) =
+    discard
+
+  proc main() =
+    var obj = Object(foo: 2)
+    obj.doSomething()
+    doAssert obj.foo == 999
+
+
+  main()

